pkg/resources/statefulsets/prometheus.go

```diff
@@ -0,0 +1,280 @@
+	BearerToken   string         `yaml:"bearer_token"`
+	MetricsPath   string         `yaml:"metrics_path"`
+	Scheme        string         `yaml:"scheme"`
+	TLSConfig     tlsConfig      `yaml:"tls_config"`
+	StaticConfigs []staticConfig `yaml:"static_configs"`
+}
+
+type prometheusConfig struct {
+	Global        globalConfig   `yaml:"global"`
+	ScrapeConfigs []scrapeConfig `yaml:"scrape_configs"`
+}
+
+func genBearerToken(accessKey, secretKey string) string {
+	jwt := jwtgo.NewWithClaims(jwtgo.SigningMethodHS512, jwtgo.StandardClaims{
+		ExpiresAt: time.Now().UTC().Add(defaultPrometheusJWTExpiry).Unix(),
+		Subject:   accessKey,
+		Issuer:    "prometheus",
+	})
+
+	token, err := jwt.SignedString([]byte(secretKey))
+	if err != nil {
+		panic(fmt.Sprintf("jwt key generation: %v", err))
+	}
+
+	return token
+}
+
+// getMinioPodAddrs returns a list of stable minio pod addresses.
+func getMinioPodAddrs(t *miniov1.Tenant) []string {
+	targets := []string{}
+	for _, pool := range t.Spec.Pools {
+		poolName := t.PoolStatefulsetName(&pool)
+		for i := 0; i < int(pool.Servers); i++ {
+			target := fmt.Sprintf("%s-%d.%s.%s.svc.%s:%d", poolName, i, t.MinIOHLServiceName(), t.Namespace, miniov1.GetClusterDomain(), miniov1.MinIOPort)
+			targets = append(targets, target)
+		}
+	}
+	return targets
+}
+
+func prometheusInitContainers(t *miniov1.Tenant, accessKey, secretKey string) []corev1.Container {
+	bearerToken := genBearerToken(accessKey, secretKey)
+	minioTargets := getMinioPodAddrs(t)
+
+	// populate config
+	promConfig := prometheusConfig{
+		Global: globalConfig{
+			ScrapeInterval:     10 * time.Second,
+			EvaluationInterval: 30 * time.Second,
+		},
+		ScrapeConfigs: []scrapeConfig{
+			{
+				JobName:     "minio",
+				BearerToken: bearerToken,
+				MetricsPath: "/minio/prometheus/metrics",
+				Scheme:      "https",
+				TLSConfig: tlsConfig{
+					CAFile: "/var/run/secrets/kubernetes.io/serviceaccount/ca.crt",
+				},
+				StaticConfigs: []staticConfig{
+					{
+						Targets: minioTargets,
+					},
+				},
+			},
+		},
+	}
+
+	d, err := yaml.Marshal(&promConfig)
+	if err != nil {
+		panic(fmt.Sprintf("error marshaling to yaml: %v", err))
+	}
+
+	scriptLines := []string{}
+	for i, line := range strings.Split(string(d), "\n") {
+		outline := ""
+		if i == 0 {
+			outline = fmt.Sprintf("echo '%s' > /etc/prometheus/prometheus.yml", line)
+		} else {
+			outline = fmt.Sprintf("echo '%s' >> /etc/prometheus/prometheus.yml", line)
+		}
+		scriptLines = append(scriptLines, outline)
+	}
+
+	scriptArg := strings.Join(scriptLines, "\n")
+
+	return []corev1.Container{
+		{
+			Name:            miniov1.PrometheusInitContainerName,
+			Image:           miniov1.PrometheusInitContainerImage,
+			ImagePullPolicy: t.Spec.ImagePullPolicy,
+			VolumeMounts:    []corev1.VolumeMount{prometheusConfigVolumeMount(t)},
+			Command:         []string{"/bin/sh"},
+			Args:            []string{"-c", scriptArg},
+			Resources:       t.Spec.Prometheus.Resources,
+		},
+	}
+}
```

Comment:
@donatello were you not able to put prometheus.yml as a configmap for the prometheus container? instead of running a InitContainer?  

this avoids running multiple containers IMO

Reply:
It is possible, but I wanted to prevent users from modifying the configmap intentionally or not. This prometheus instance is only meant for our simple/opinionated metrics visualization in the console, so I didn't want to expose it.

Reply:
Even then that is only possible with the right service accounts and priviledges - obscuring that fact makes it harder to deploy changes to prometheus.yml IMO. 

Even if we do a simpler approach a better declarative style is configMap and it's easier to manage. Basically minio-operator knows the state of prometheus configmap, if its modifed it can react and change it if necessary - these are the functionalities that are available won't be possible with initContainer style approach.